modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/leases/LeaseBatchSerializer.java

```diff
@@ -0,0 +1,485 @@
+        return dictionary.nameCount() <= MAX_NODES_FOR_COMPACT_MODE;
+    }
+
+    private static int flags(
+            boolean accepted,
+            boolean prolongable,
+            boolean hasProposedCandidate,
+            boolean hasUncommonPeriod,
+            boolean hasExpirationLogicalPart
+    ) {
+        return (accepted ? ACCEPTED_MASK : 0)
+                | (prolongable ? PROLONGABLE_MASK : 0)
+                | (hasProposedCandidate ? HAS_PROPOSED_CANDIDATE_MASK : 0)
+                | (hasUncommonPeriod ? HAS_UNCOMMON_PERIOD_MASK : 0)
+                | (hasExpirationLogicalPart ? HAS_EXPIRATION_LOGICAL_PART_MASK : 0);
+    }
+
+    @Override
+    protected LeaseBatch readExternalData(byte protoVer, IgniteDataInput in) throws IOException {
+        long minTimestampPhysical = in.readVarInt();
+        long commonLeasePeriod = in.readVarInt();
+        NodesDictionary nodesDictionary = NodesDictionary.readFrom(in);
+
+        List<Lease> leases = new ArrayList<>();
+
+        readPartitionedGroupLeases(minTimestampPhysical, commonLeasePeriod, nodesDictionary, leases, in, TablePartitionId::new);
+
+        return new LeaseBatch(leases);
+    }
+
+    private static void readPartitionedGroupLeases(
+            long minTimestampPhysical,
+            long commonLeasePeriod,
+            NodesDictionary nodesDictionary,
+            List<Lease> leases,
+            IgniteDataInput in,
+            GroupIdFactory groupIdFactory
+    ) throws IOException {
+        int objectCount = in.readVarIntAsInt();
+
+        int objectIdBase = 0;
+        for (int i = 0; i < objectCount; i++) {
+            objectIdBase = readLeasesForObject(
+                    minTimestampPhysical,
+                    commonLeasePeriod,
+                    nodesDictionary,
+                    leases,
+                    in,
+                    groupIdFactory,
+                    objectIdBase
+            );
+        }
+    }
+
+    private static int readLeasesForObject(
+            long minTimestampPhysical,
+            long commonLeasePeriod,
+            NodesDictionary nodesDictionary,
+            List<Lease> leases,
+            IgniteDataInput in,
+            GroupIdFactory groupIdFactory,
+            int objectIdBase
+    ) throws IOException {
+        int objectId = objectIdBase + in.readVarIntAsInt();
+
+        int partitionCount = in.readVarIntAsInt();
+        for (int partitionId = 0; partitionId < partitionCount; partitionId++) {
+            Lease lease = readLeaseForPartition(
+                    partitionId,
+                    objectId,
+                    minTimestampPhysical,
+                    commonLeasePeriod,
+                    nodesDictionary,
+                    in,
+                    groupIdFactory
+            );
+            if (lease != null) {
+                leases.add(lease);
+            }
+        }
+
+        return objectId;
+    }
+
+    private static @Nullable Lease readLeaseForPartition(
```

Comment:
could you please describe in text (or better ascii-scheme) the data input structure and how the lease is stored? You may also do this in class header

Reply:
Added the format grammar